lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+
+	return hostname, nil
+}
+
+// getVolumeType returns the selected provisioning type of the volume.
+func (d *powerflex) getVolumeType(vol Volume) string {
+	if vol.config["block.type"] != "" {
+		return vol.config["block.type"]
+	}
+
+	return "thin"
+}
+
+// getVolumeName returns the fully qualified name derived from the volume.
+func (d *powerflex) getVolumeName(vol Volume, snapName string) string {
+	var out string
+	parentName, snapshotName, isSnapshot := api.GetParentAndSnapshotName(vol.name)
+
+	// Only use filesystem suffix on filesystem type image volumes (for all content types).
+	// Since PowerFlex volumes can be 31 characters long at max, use the image fingerprint.
+	if vol.volType == VolumeTypeImage {
+		parentName = fmt.Sprintf("%s_%s", parentName[0:12], vol.ConfigBlockFilesystem())
+	}
+
+	if vol.contentType == ContentTypeBlock {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexBlockVolSuffix)
+	} else if vol.contentType == ContentTypeISO {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexISOVolSuffix)
+	}
+
+	// Use volume's type as storage volume prefix, unless there is an override in powerFlexVolTypePrefixes.
+	volumeTypePrefix := string(vol.volType)
+	volumeTypePrefixOverride, foundOveride := powerFlexVolTypePrefixes[vol.volType]
+	if foundOveride {
+		volumeTypePrefix = volumeTypePrefixOverride
+	}
+
+	if snapName != "" {
+		// Always use the provided snapshot name if specified.
+		out = fmt.Sprintf("%s_%s@%s", volumeTypePrefix, parentName, snapName)
+	} else {
+		if isSnapshot {
+			// If volumeName is a snapshot (<vol>/<snap>) and snapName is not set,
+			// assume that it's a normal snapshot (not a zombie) and prefix it with
+			// "snapshot_".
+			out = fmt.Sprintf("%s_%s@snapshot_%s", volumeTypePrefix, parentName, snapshotName)
+		} else {
+			out = fmt.Sprintf("%s_%s", volumeTypePrefix, parentName)
+		}
+	}
+
+	return out
+}
+
+// createNVMeHost creates this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) createNVMeHost() (string, error) {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return "", err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	var hostID string
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		if !api.StatusErrorCheck(err, http.StatusNotFound) {
+			return "", err
+		}
+
+		hostname, err := d.getHostname()
+		if err != nil {
+			return "", err
+		}
+
+		hostID, err = d.powerFlex().createHost(hostname, nqn)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if hostID == "" {
+		hostID = host.ID
+	}
+
+	return hostID, nil
+}
+
+// deleteNVMeHost delets this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) deleteNVMeHost() error {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
```

Comment:
This pattern of using `unlock` without checking if its nil will eventually cause a LXD server crash, as if the `locking.Lock` call is being blocked because another process has the lock and then d.state.ShutdownCtx is cancelled then nil will be returned, which will then be called when the function ends and crash the server.

Also we should only proceed with the action if a non-nil unlock was returned as otherwise we've not got the lock.

Reply:
To be honest I just copied the pattern from another place but it had `context.TODO()` in there which won't get cancelled. Makes total sense, thanks.

Reply:
maybe we should change locking.Lock to return an error as well, that way it'd be more conventional and clear that if err != nil then the unlock function is nil.

Reply:
Let me add an additional PR for this. In most places `locking.Lock` is used with `context.Background()` or `context.TODO()` but there are some that might also produce a crash.

Reply:
Rebased from `main` and the error is now handled appropriately.